README
Food-ECodes

The README  is  used  to  introduce the module and provide instructions on how to
install the module, any machine dependencies it may have (for example C compilers
and installed libraries) and any other information that should be provided before
the module is installed.

A  README  file  is required for CPAN modules since CPAN extracts the README file
from a module distribution so that people browsing the archive can use it to  get
an  idea  of  the  module's  uses.  It  is usually a good idea to provide version
information here so that people can decide whether fixes for the module are worth
downloading.

INSTALLATION

To install this module, run the following commands:

    perl Makefile.PL
    make
    make test
    make install

SUPPORT AND DOCUMENTATION

After  installing,   you can  find documentation for this module with the perldoc
command.

    perldoc Food::ECodes

You can also look for information at:

    RT, CPAN's request tracker
        http://rt.cpan.org/NoAuth/Bugs.html?Dist=Food-ECodes

    AnnoCPAN, Annotated CPAN documentation
        http://annocpan.org/dist/Food-ECodes

    CPAN Ratings
        http://cpanratings.perl.org/d/Food-ECodes

    Search CPAN
        http://search.cpan.org/dist/Food-ECodes/

LICENSE AND COPYRIGHT

Copyright (C) 2011 - 2015 Mohammad S Anwar

This program is free software; you can redistribute it and/or modify it under the
terms of the the Artistic License (2.0).You may obtain a copy of the full license
at:

L<http://www.perlfoundation.org/artistic_license_2_0>

Any  use,  modification, and distribution of the Standard or Modified Versions is
governed by this Artistic License.By using, modifying / distributing the Package,
you accept this license. Do not use, modify, or distribute the Package, if you do
not accept this license.

If your Modified Version has been derived from a Modified Version made by someone
other than you, you're nevertheless required to ensure that your Modified Version
complies with the requirements of this license.

This  license  does  not  grant you the right to use any trademark, service mark,
tradename, or logo of the Copyright Holder.

This license includes the non-exclusive, worldwide, free-of-charge patent license
to  make,  have made, use, offer to sell, sell, import and otherwise transfer the
Package with respect to any patent claims licensable by the Copyright Holder that
are  necessarily  infringed  by the  Package.  If you institute patent litigation
(including  a  cross-claim or  counterclaim)  against any party alleging that the
Package constitutes direct or contributory patent infringement,then this Artistic
License to you shall terminate on the date that such litigation is filed.

Disclaimer  of  Warranty:  THE  PACKAGE  IS  PROVIDED BY THE COPYRIGHT HOLDER AND
CONTRIBUTORS  "AS IS'  AND WITHOUT ANY EXPRESS OR IMPLIED WARRANTIES. THE IMPLIED
WARRANTIES    OF     MERCHANTABILITY,   FITNESS   FOR   A PARTICULAR  PURPOSE, OR
NON-INFRINGEMENT ARE DISCLAIMED TO THE EXTENT PERMITTED BY YOUR LOCAL LAW. UNLESS
REQUIRED BY LAW, NO COPYRIGHT HOLDER OR CONTRIBUTOR WILL BE LIABLE FOR ANY DIRECT,
INDIRECT, INCIDENTAL, OR CONSEQUENTIAL DAMAGES ARISING IN ANY WAY OUT OF THE USE
OF THE PACKAGE,EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

Changes
Revision history for Food::ECodes

0.16  Wed Sep 07 09:15:00 2016
      - Added test scripts (t/meta-json.t and t/meta-yml.t).

0.15  Sat Aug 13 15:55:00 2016
      - Updated Copyright year information.
      - Tidied up github repository url.
      - Tidied up Makefile.PL script.

0.14  Tue Dec 15 10:40:00 2015
      - Added key 'provides' to the Makefile.PL script.

0.13  Mon Aug 03 12:00:00 2015
      - Added section "DESCRIPTION" and "SYNOPSIS" to the pod document.

0.12  Tue Jun 02 11:40:00 2015
      - Added LICENSE file.

0.11  Sat Jan 17 00:40:00 2015
      - Removed SIGNATURE file.
      - Updated MANIFEST.SKIP file to ignore folder 'eumm/'.
      - Updated Copyright year information.

0.10  Sun Sep 21 02:21:34 2014
      - Added method purpose() to the Food::ECodes::Additive package.

0.09  Sun Sep 21 01:51:05 2014
      - Re-structured the code using 'Moo'.
      - Added placeholder for 'Additive'.
      - Updated the interface of Food::ECodes completely.
      - Stored the ecodes definition in XML file, added to the distribution.
      - Removed dependency on 'Readonly' and 'Carp'.

0.08  Tue Aug 09 15:44:11 2011
      - Made Readonly v1.03 as minimum requirement.
      - Prepared the distribution using the Module::Signature v0.68.

0.07  Wed May 25 10:21:00 2011
      - Added E441 to the list of Non-Veg ECode [https://rt.cpan.org/Ticket/Display.html?id=68455].

0.06  Tue Apr 26 10:07:00 2011
      - Added method is_vegetarian() as suggested by Sawyer.
      - Added test case for method is_vegetarian().

0.05  Thu Apr 21 18:36:23 2011
      - Renamed title "SYNOPSIS" to "DESCRIPTION".
      - Minor cosmetics changes.

0.04  Mon Apr 11 13:49:55 2011
      - Fixed the typo in the pod for method get_status().
      - Fixed the typo in the pod for method get_name().
      - Fixed the typo in the pod for method get_purpose().
      - Fixed the syntax error in the example pod for method is_valid().
      - Fixed the syntax error in the example pod for method is_eu_approved().
      - Fixed the syntax error in the example pod for method is_us_approved().

0.03  Mon Apr 11 12:33:18 2011
      - Added remaining ECodes from 1000 upto 1600.
      - Added method get_status().
      - Updated unit test 01-unit-test.t for new method get_status().

0.02  Fri Apr 08 16:42:22 2011
      - Fixed the typo in the pod for method is_valid().
      - Fixed the typo in the pod for method is_eu_approved().
      - Fixed the typo in the pod for method is_us_approved().

0.01  Fri Apr 08 11:00:00 2011
      - First version released.

Makefile.PL
use 5.006;
use strict;
use warnings FATAL => 'all';
use ExtUtils::MakeMaker;
use File::ShareDir::Install;

install_share dist => 'share';

WriteMakefile(
    NAME               => 'Food::ECodes',
    AUTHOR             => q{Mohammad S Anwar <mohammad.anwar@yahoo.com>},
    VERSION_FROM       => 'lib/Food/ECodes.pm',
    ABSTRACT_FROM      => 'lib/Food/ECodes.pm',
    LICENSE            => 'artistic_2',
    MIN_PERL_VERSION   => 5.006,
    CONFIGURE_REQUIRES => {
        'ExtUtils::MakeMaker'     => 0,
        'File::ShareDir::Install' => 0,
    },
    BUILD_REQUIRES   => {
        'Test::More' => 0,
    },
    PREREQ_PM => {
        'Moo'              => '2.000000',
        'namespace::clean' => '0.24',
        'XML::Simple'      => '2.20',
        'File::Share'      => '0.25',
    },
    dist  => { COMPRESS => 'gzip -9f', SUFFIX => 'gz', },
    clean => { FILES => 'Food-ECodes-*' },
    (eval { ExtUtils::MakeMaker->VERSION(6.46) } ? (META_MERGE => {
        'meta-spec' => { version => 2 },
        provides    => {
            'Food::ECodes'           => { file => 'lib/Food/ECodes.pm'         , version => '0.16' },
            'Food::ECodes::Additive' => { file => 'lib/Food/ECodes/Additive.pm', version => '0.16' },
        },
        resources => {
            repository => {
                type => 'git',
                url  => 'https://github.com/manwar/Food-ECodes.git',
                web  => 'https://github.com/manwar/Food-ECodes',
            },
        }})
     : ()
    ),
);


package MY;
use File::ShareDir::Install 'postamble';


lib/Food/ECodes.pm
package Food::ECodes;

$Food::ECodes::VERSION = '0.16';

=head1 NAME

Food::ECodes - Interface to Food Additive ECodes.

=head1 VERSION

Version 0.16

=cut

use 5.006;
use XML::Simple;
use Data::Dumper;
use File::Share ':all';
use Food::ECodes::Additive;

use Moo;
use namespace::clean;

has 'additives' => (is => 'ro');

sub BUILDARGS {
    my ($class, $args) = @_;

    die "ERROR: No parameters required for constructor." if defined $args;

    my $xml       = dist_file('Food-ECodes', 'ecodes.xml');
    my $ecode     = XMLin($xml, ForceArray => 1, KeyAttr => [ 'ecode' ]);
    my $additives = {};

    foreach my $_ecode (@{$ecode->{ecode}}) {
        $additives->{uc($_ecode->{id})} = Food::ECodes::Additive->new($_ecode);
    }

    return { additives => $additives };
};

=head1 DESCRIPTION

It provides simple interface to the Ecodes library.

=head1 SYNOPSIS

    use strict; use warnings;
    use Food::ECodes;

    my $ecode    = Food::ECodes->new;
    my $additive = $ecode->search('E100');
    print "Name: ", $additive->name, "\n";

=head1 METHODS

=head2 search()

Returns  object  of  type L<Food::ECodes::Additive> on success, which can be then
further queried.

=cut

sub search {
    my ($self, $code) = @_;

    die "ERROR: Missing parameter 'ecode'."      unless defined $code;
    die "ERROR: Invalid ecode '$code' received." unless (exists $self->{additives}->{uc($code)});

    return $self->{additives}->{$code};
}

=head1 AUTHOR

Mohammad S Anwar, C<< <mohammad.anwar at yahoo.com> >>

=head1 REPOSITORY

L<https://github.com/manwar/Food-ECodes>

=head1 BUGS

Please report any bugs or feature requests  to  C<bug-food-ecodes at rt.cpan.org>
or through the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Food-ECodes>.
I will be notified, and then you'll automatically be notified of progress on your
bug as I make changes.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc Food::ECodes

You can also look for information at:

=over 4

=item * RT: CPAN's request tracker

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=Food-ECodes>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/Food-ECodes>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/Food-ECodes>

=item * Search CPAN

L<http://search.cpan.org/dist/Food-ECodes/>

=back

=head1 LICENSE AND COPYRIGHT

Copyright (C) 2011 - 2016 Mohammad S Anwar.

This  program  is  free software; you can redistribute it and/or modify it under
the  terms  of the the Artistic License (2.0). You may obtain a copy of the full
license at:

L<http://www.perlfoundation.org/artistic_license_2_0>

Any  use,  modification, and distribution of the Standard or Modified Versions is
governed by this Artistic License.By using, modifying or distributing the Package,
you accept this license. Do not use, modify, or distribute the Package, if you do
not accept this license.

If your Modified Version has been derived from a Modified Version made by someone
other than you,you are nevertheless required to ensure that your Modified Version
 complies with the requirements of this license.

This  license  does  not grant you the right to use any trademark,  service mark,
tradename, or logo of the Copyright Holder.

This license includes the non-exclusive, worldwide, free-of-charge patent license
to make,  have made, use,  offer to sell, sell, import and otherwise transfer the
Package with respect to any patent claims licensable by the Copyright Holder that
are  necessarily  infringed  by  the  Package. If you institute patent litigation
(including  a  cross-claim  or  counterclaim) against any party alleging that the
Package constitutes direct or contributory patent infringement,then this Artistic
License to you shall terminate on the date that such litigation is filed.

Disclaimer  of  Warranty:  THE  PACKAGE  IS  PROVIDED BY THE COPYRIGHT HOLDER AND
CONTRIBUTORS  "AS IS'  AND WITHOUT ANY EXPRESS OR IMPLIED WARRANTIES. THE IMPLIED
WARRANTIES    OF   MERCHANTABILITY,   FITNESS   FOR   A   PARTICULAR  PURPOSE, OR
NON-INFRINGEMENT ARE DISCLAIMED TO THE EXTENT PERMITTED BY YOUR LOCAL LAW. UNLESS
REQUIRED BY LAW, NO COPYRIGHT HOLDER OR CONTRIBUTOR WILL BE LIABLE FOR ANY DIRECT,
INDIRECT, INCIDENTAL,  OR CONSEQUENTIAL DAMAGES ARISING IN ANY WAY OUT OF THE USE
OF THE PACKAGE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

=cut

1; # End of Food::ECodes


lib/Food/ECodes/Additive.pm
package Food::ECodes::Additive;

$Food::ECodes::Additive::VERSION = '0.16';

=head1 NAME

Food::ECodes::Additive - Placeholder for Food Additive.

=head1 VERSION

Version 0.16

=cut

use 5.006;
use Data::Dumper;

use Moo;
use namespace::clean;

has 'id'          => (is => 'ro');
has 'name'        => (is => 'ro');
has 'status'      => (is => 'ro');
has 'purpose'     => (is => 'ro');
has 'eu_approved' => (is => 'ro');
has 'us_approved' => (is => 'ro');
has 'vegetarian'  => (is => 'ro');

=head1 SYNOPSIS

    use strict; use warnings;
    use Food::ECodes;

    my $ecode    = Food::ECodes->new;
    my $additive = $ecode->search('E100');

    print "Id         : ", $additive->id,          "\n";
    print "Name       : ", $additive->name,        "\n";
    print "Status     : ", $additive->status,      "\n";
    print "Purpose    : ", $additive->purpose,     "\n";
    print "EU Approved: ", $additive->eu_approved, "\n";
    print "US Approved: ", $additive->us_approved, "\n";
    print "Vegetarian : ", $additive->vegetarian,  "\n";

=head1 METHODS

=head2 id()

=head2 name()

=head2 status()

=head2 purpose()

=head2 eu_approved()

=head2 us_approved()

=head2 vegetarian()

=head1 AUTHOR

Mohammad S Anwar, C<< <mohammad.anwar at yahoo.com> >>

=head1 REPOSITORY

L<https://github.com/manwar/Food-ECodes>

=head1 BUGS

Please report any bugs or feature requests  to  C<bug-food-ecodes at rt.cpan.org>
or through the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Food-ECodes>.
I will be notified, and then you'll automatically be notified of progress on your
bug as I make changes.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc Food::ECodes::Additive

You can also look for information at:

=over 4

=item * RT: CPAN's request tracker

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=Food-ECodes>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/Food-ECodes>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/Food-ECodes>

=item * Search CPAN

L<http://search.cpan.org/dist/Food-ECodes/>

=back

=head1 LICENSE AND COPYRIGHT

Copyright (C) 2011 - 2016 Mohammad S Anwar.

This  program  is  free software; you can redistribute it and/or modify it under
the  terms  of the the Artistic License (2.0). You may obtain a copy of the full
license at:

L<http://www.perlfoundation.org/artistic_license_2_0>

Any  use,  modification, and distribution of the Standard or Modified Versions is
governed by this Artistic License.By using, modifying or distributing the Package,
you accept this license. Do not use, modify, or distribute the Package, if you do
not accept this license.

If your Modified Version has been derived from a Modified Version made by someone
other than you,you are nevertheless required to ensure that your Modified Version
 complies with the requirements of this license.

This  license  does  not grant you the right to use any trademark,  service mark,
tradename, or logo of the Copyright Holder.

This license includes the non-exclusive, worldwide, free-of-charge patent license
to make,  have made, use,  offer to sell, sell, import and otherwise transfer the
Package with respect to any patent claims licensable by the Copyright Holder that
are  necessarily  infringed  by  the  Package. If you institute patent litigation
(including  a  cross-claim  or  counterclaim) against any party alleging that the
Package constitutes direct or contributory patent infringement,then this Artistic
License to you shall terminate on the date that such litigation is filed.

Disclaimer  of  Warranty:  THE  PACKAGE  IS  PROVIDED BY THE COPYRIGHT HOLDER AND
CONTRIBUTORS  "AS IS'  AND WITHOUT ANY EXPRESS OR IMPLIED WARRANTIES. THE IMPLIED
WARRANTIES    OF   MERCHANTABILITY,   FITNESS   FOR   A   PARTICULAR  PURPOSE, OR
NON-INFRINGEMENT ARE DISCLAIMED TO THE EXTENT PERMITTED BY YOUR LOCAL LAW. UNLESS
REQUIRED BY LAW, NO COPYRIGHT HOLDER OR CONTRIBUTOR WILL BE LIABLE FOR ANY DIRECT,
INDIRECT, INCIDENTAL,  OR CONSEQUENTIAL DAMAGES ARISING IN ANY WAY OUT OF THE USE
OF THE PACKAGE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

=cut

1; # End of Food::ECodes::Additive
